### PR TITLE
allow forking VMs by refcounting dictionaries

### PR DESCRIPTION
**Pull request by hawkw**

### CODE REVIEW

File: src/vm/async_vm.rs [hunk 1]
```diff
@@ -57,17 +57,34 @@ where
...
         async_builtins: A,
     ) -> Result<Self, Error> {
-        let vm = Forth::new_async(dstack_buf, rstack_buf, cstack_buf, dict_buf, input, output, host_ctxt, sync_builtins, A::BUILTINS)?;
+        let vm = Forth::new_async(dstack_buf, rstack_buf, cstack_buf, dict, input, output, host_ctxt, sync_builtins, A::BUILTINS)?;
         Ok(Self { vm, builtins: async_builtins })
     }
 
+    pub unsafe fn fork(
+        &mut self,
+        mut new_dict: OwnedDict<T>,
```
Comment (jamesmunns) [inline src/vm/async_vm.rs:73]:
```suggestion
        new_dict: OwnedDict<T>,
```

File: src/vm/builtins.rs [hunk 2]
```diff
@@ -341,23 +341,37 @@ impl<T: 'static> Forth<T> {
...
+                unsafe {
+                    name_ptr.write_bytes(0x00, len);
+                }
+                self.dict.alloc.cur = name_ptr;
+            },
+            // The definition is in a parent (frozen) dictionary. We can't
+            // mutate that dictionary, so we must create a new entry in the
+            // current dict saying that the definition is forgotten.
+            // XXX(eliza): or this could be a runtime error? IDK...
+            DictLocation::Parent(de) => {
```
Comment (jamesmunns) [inline src/vm/builtins.rs:370]:
```suggestion
            DictLocation::Parent(_de) => {
```

**Review by jamesmunns — APPROVED** [review]
A couple of nits, let's get to using it :D

File: src/vm/builtins.rs [hunk 3]
```diff
@@ -225,19 +225,19 @@ impl<T: 'static> Forth<T> {
...
-            output.write_str(item.hdr.name.as_str())?;
+        for item in dict.entries() {
+            output.write_str(unsafe { item.entry().as_ref() }.hdr.name.as_str())?;
+            if let DictLocation::Parent(_) = item {
+                // indicate that this binding is inherited from a parent.
+                // XXX(eliza): i was initially gonna add "(inherited)" but that
+                // makes some of the tests overflow their output buffer, lol.
+                // idk if the output of the `dict` word is standardized, is this
+                // okay to do?
+                output.write_str("*")?;
```
Comment (hawkw) [inline src/vm/builtins.rs:238]: @jamesmunns do you have an opinion on this?

Comment (jamesmunns) [reply src/vm/builtins.rs:238]: Nope! This is just a purely invented-by-me debugging util. No format stability.

**Pull request merged**